flink-yarn/src/main/java/org/apache/flink/yarn/YarnClusterDescriptor.java

```diff
@@ -1568,151 +1568,78 @@ public static ApplicationSubmissionContextReflector getInstance() {
-                        ATTEMPT_FAILURES_METHOD_NAME);
-            } catch (NoSuchMethodException e) {
-                LOG.debug(
-                        "{} does not support method {}.",
-                        clazz.getCanonicalName(),
-                        ATTEMPT_FAILURES_METHOD_NAME);
-                // assign null because the Hadoop version apparently does not support this call.
-                attemptFailuresValidityIntervalMethod = null;
-            }
-
-            this.attemptFailuresValidityIntervalMethod = attemptFailuresValidityIntervalMethod;
-
-            try {
-                // this method is only supported by Hadoop 2.4.0 onwards
-                keepContainersMethod = clazz.getMethod(KEEP_CONTAINERS_METHOD_NAME, boolean.class);
-                LOG.debug(
-                        "{} supports method {}.",
-                        clazz.getCanonicalName(),
-                        KEEP_CONTAINERS_METHOD_NAME);
-            } catch (NoSuchMethodException e) {
-                LOG.debug(
-                        "{} does not support method {}.",
-                        clazz.getCanonicalName(),
-                        KEEP_CONTAINERS_METHOD_NAME);
-                // assign null because the Hadoop version apparently does not support this call.
-                keepContainersMethod = null;
-            }
-
-            this.keepContainersMethod = keepContainersMethod;
+            this.applicationTagsMethod = getMethod(clazz, APPLICATION_TAGS_METHOD_NAME, Set.class);
+            this.attemptFailuresValidityIntervalMethod =
+                    getMethod(clazz, ATTEMPT_FAILURES_METHOD_NAME, long.class);
+            this.keepContainersMethod =
+                    getMethod(clazz, KEEP_CONTAINERS_METHOD_NAME, boolean.class);
+            this.nodeLabelExpressionMethod =
+                    getMethod(clazz, NODE_LABEL_EXPRESSION_NAME, String.class);
+        }
 
+        private Method getMethod(Class<?> clazz, String methodName, Class<?>... paramTypes) {
             try {
-                nodeLabelExpressionMethod =
-                        clazz.getMethod(NODE_LABEL_EXPRESSION_NAME, String.class);
-                LOG.debug(
-                        "{} supports method {}.",
-                        clazz.getCanonicalName(),
-                        NODE_LABEL_EXPRESSION_NAME);
+                Method method = clazz.getMethod(methodName, paramTypes);
+                LOG.debug("{} supports method {}.", clazz.getCanonicalName(), methodName);
+                return method;
             } catch (NoSuchMethodException e) {
-                LOG.debug(
-                        "{} does not support method {}.",
-                        clazz.getCanonicalName(),
-                        NODE_LABEL_EXPRESSION_NAME);
-                nodeLabelExpressionMethod = null;
+                LOG.debug("{} does not support method {}.", clazz.getCanonicalName(), methodName);
+                return null;
```

Comment:
// assign null because the Hadoop version apparently does not support this call.